src/editor/editor_brush.cpp

```diff
@@ -0,0 +1,409 @@
+}
+
+auto& CBrush::getDecoration(const TDecorationOptions &options) {
+	static SDecoration emptyValue{};
+	if (!decorationsPalette.count(options)) {
+		const auto prevSize = decorationsPalette.size();
+		generateDecoration();
+		if (decorationsPalette.size() == prevSize) {
+			ErrorPrint("Unable to generate decoration with current set of options\n");
+			return emptyValue;
+		}
+	}
+	return decorationsPalette[options];
+}
+
+void CBrush::pushDecorationTiles(uint8_t srcWidth, uint8_t srcHeight, const std::vector<tile_index> &srcTiles)
+{
+	if (srcWidth * srcHeight != srcTiles.size()) {
+		return;
+	}
+	decorationsPalette[decorationOptions] = {srcWidth, srcHeight, srcTiles};
+}
+
+void CBrush::loadDecoration()
+{
+	const auto &[srcWidth, srcHeight, srcTiles] = getDecoration(decorationOptions);
+	setTiles(srcWidth, srcHeight, srcTiles);
+}
+
+void CBrush::generateDecoration()
+{
+	const fs::path filename = LibraryFileName(properties.decorationGenerator.source);
+
+	if (LuaLoadFile(filename) == -1) {
+		ErrorPrint("%s's brush generator file '%s' not found\n",
+					name.c_str(),
+					properties.decorationGenerator.source.c_str());
+	}
+}
+
+tile_index CBrush::getCurrentTile() const
+{
+	for (const auto &tile : tiles) {
+		if (tile) {
+			return tile;
+		}
+	}
+	return 0;
+}
+
+// Bresenham algorithm
+void CBrush::drawCircle(int16_t xCenter,
+						int16_t yCenter,
+						int16_t diameter,
+						tile_index tile,
+						std::vector<tile_index> &canvas)
+{
+	// Because of the symmetry with respect to the cursor position, the diameter must be odd
+	if (canvas.size() < diameter * diameter || diameter % 2 == 0) {
+		return;
+	}
+
+	auto drawHLine = [diameter, tile, &canvas](int16_t x1, int16_t x2, int16_t y)
+	{
+		const size_t idx = y * diameter;
+		std::fill(&canvas[idx + x1], &canvas[idx + x2 + 1], tile);
+	};
+
+	int x = diameter / 2;
+	int y = 0;
+	int delta = 1 - x;
+
+	while (x >= y)
+	{
+		drawHLine(xCenter - x, xCenter + x, yCenter + y);
+		drawHLine(xCenter - x, xCenter + x, yCenter - y);
+		drawHLine(xCenter - y, xCenter + y, yCenter + x);
+		drawHLine(xCenter - y, xCenter + y, yCenter - x);
+
+		y++;
+		if (delta < 0)
+		{
+			delta += 2 * y + 1;
+		}
+		else
+		{
+			x--;
+			delta += 2 * (y - x + 1);
+		}
+	}
+}
+
+void CBrushesSet::loadBrushes(std::string_view brushesSrc /* = {} */)
+{
+	brushes.clear();
+	if (!brushesSrc.empty()) {
+		this->brushesSrc = brushesSrc;
+		const fs::path filename = LibraryFileName(this->brushesSrc);
+		if (LuaLoadFile(filename) == -1) {
+			ErrorPrint("Load failed: \"%s\"\n", filename.u8string().c_str());
+			this->brushesSrc = {};
+		}
+	}
+	if (brushes.empty()) {
+		brushes.push_back(CBrush(std::string("Default"), CBrush::Properties()));
+	}
+	setCurrentBrush(brushes.begin()->getName());
+}
+
+bool CBrushesSet::setCurrentBrush(std::string_view name)
+{
+	Assert(!brushes.empty());
+
+	const auto prev = currentBrush.getName();
+	for (auto &brush : brushes) {
+		if (brush.getName() == name) {
```

Comment:
with your final `break`, I tend to use `auto it = ranges::find_if` in that case.

Reply:
fixed:
https://github.com/Wargus/stratagus/pull/732/commits/8b20c110b31bd45a35ddfa08c137c6a1eff7be67